### PR TITLE
Fixing race condition between roles after doing handshake

### PR DESCRIPTION
**Pull request by darricksee**
Currently there is a race condition after doing the handshake between upstream/downstream roles. Here's the situation roughly:

1. Pool sends final handshake message
2. Pool loops and waits for the message to go out
3. Proxy receives final handshake message 
4. Proxy immediately sends SetupConnection message
5. Pool receives SetupConnection message
6. Pool hasn't yet transitioned from Handshake mode to Transport mode so it thinks the incoming message is another Handshake message and errors out.

This fix removes step 2 in favor of the following flow

1. Pool encodes final handshake message
2. Pool transitions from Handshake mode to Transport mode
3. Pool sends final handshake message
4. Proxy receives final handshake message
5. Proxy immediately send SetupConnection message
6. Pool receives SetupConnection message
7. Pool is already in Transport mode so properly decrypts message and handles it.

### CODE REVIEW

File: roles/v2/mining-proxy/src/lib/upstream_mining.rs [hunk 1]
```diff
@@ -414,15 +414,25 @@ impl UpstreamMiningNode {
...
         sv2_frame: StdFrame,
     ) -> Result<(), crate::lib::error::Error> {
-        let (has_sv2_connetcion, mut connection) = self_mutex
-            .safe_lock(|self_| (self_.sv2_connection.is_some(), self_.connection.clone()))
+        let (has_sv2_connection, mut connection, address) = self_mutex
+            .safe_lock(|self_| {
+                (
+                    self_.sv2_connection.is_some(),
+                    self_.connection.clone(),
+                    self_.address,
```
Comment (darricksee) [inline roles/v2/mining-proxy/src/lib/upstream_mining.rs:422]: Just using address for logging purposes

File: roles/v2/pool/src/lib/mining_pool/mod.rs [hunk 2]
```diff
@@ -60,10 +60,12 @@ impl Downstream {
         pool: Arc<Mutex<Pool>>,
         channel_factory: Arc<Mutex<PoolChannelFactory>>,
         status_tx: status::Sender,
+        address: SocketAddr,
```
Comment (darricksee) [inline roles/v2/pool/src/lib/mining_pool/mod.rs:63]: Passing address through for logging purposes.

File: roles/v2/pool/src/lib/mining_pool/mod.rs [hunk 3]
```diff
@@ -233,6 +236,7 @@ impl Pool {
             config.listen_address
         );
         while let Ok((stream, _)) = listener.accept().await {
+            let address = stream.peer_addr().unwrap();
```
Comment (coleFD) [inline roles/v2/pool/src/lib/mining_pool/mod.rs:239]: is it safe to unwrap this?

File: utils/network-helpers/src/noise_connection_tokio.rs [hunk 4]
```diff
@@ -31,6 +31,8 @@ impl Connection {
         Receiver<StandardEitherFrame<Message>>,
         Sender<StandardEitherFrame<Message>>,
     ) {
+        let address = stream.peer_addr().unwrap();
```
Comment (coleFD) [inline utils/network-helpers/src/noise_connection_tokio.rs:34]: same as above

**Review by Fi3 — APPROVED** [review]

File: roles/v2/pool/src/lib/mining_pool/mod.rs (hunk 3 above)
Comment (darricksee) [reply roles/v2/pool/src/lib/mining_pool/mod.rs:239]: yeah - it looks safe. In the example code they are using `?` - I followed the call down to the bottom and it looks safe to me.

File: utils/network-helpers/src/noise_connection_tokio.rs (hunk 4 above)
Comment (darricksee) [reply utils/network-helpers/src/noise_connection_tokio.rs:34]: same.

**Pull request merged**